File: static/java/DictTable.java
// ==========================================
// MyAirtable static runtime.
// ==========================================

package myairtable;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

/**
 * Raw-dict record accessor for an Airtable table. Forwards all I/O to the {@link AirtableClient}.
 * Values flow through as {@link Fields} (dual ID/name-keyed {@code Map<String, JsonNode>}),
 * bypassing typed {@code {Table}Model} decoding.
 *
 * <p>Paired with {@code OrmTable<T>} (J-F4) — both share the same underlying client. Generated
 * {@code {Table}Table.dict()} exposes this accessor.
 */
public final class DictTable {

  /** Airtable's hard cap for create / update / delete batches. */
  public static final int BATCH_SIZE = 10;

  /**
   * Max in-flight requests for the multi-ID {@link #get(List)} fan-out. Kept just under Airtable's
   * 5-requests-per-second limit so large ID lists don't trigger a synchronized 429 storm.
   */
  public static final int MAX_CONCURRENT_GETS = 4;

  /** A fully-loaded record: id + createdTime + {@link Fields}. */
  public record Record(String id, Instant createdTime, Fields fields) {}

  // ---- Wire envelopes ----

  record RawRecordEnvelope(
      @JsonProperty("id") String id,
      @JsonProperty("createdTime") Instant createdTime,
      @JsonProperty("fields") Map<String, JsonNode> fields) {}

  record RawListResponse(
      @JsonProperty("records") List<RawRecordEnvelope> records,
      @JsonProperty("offset") String offset) {}

  /** A single update: record ID + the fields to patch. */
  public record Update(String id, Fields fields) {}

  private final String tableId;
  private final Map<String, String> nameToId;
  private final AirtableClient client;

  public DictTable(String tableId, Map<String, String> nameToId, AirtableClient client) {
    this.tableId = tableId;
    this.nameToId = nameToId;
    this.client = client;
  }

  public String getTableId() {
    return tableId;
  }

  private Record toRecord(RawRecordEnvelope raw) {
    Fields fields = new Fields(raw.fields() == null ? Map.of() : raw.fields(), nameToId);
    return new Record(raw.id(), raw.createdTime(), fields);
  }

  private static <T> T decode(String payload, Class<T> type) {
    try {
      return AirtableJson.MAPPER.readValue(payload, type);
    } catch (AirtableException e) {
      throw e;
    } catch (Exception e) {
      throw new AirtableException.Decoding(e);
    }
  }

  // ---- get (read) ----

  /** Fetch one record by ID. */
  public Record get(String recordId) {
    String payload = client.getRecord(tableId, recordId);
    return toRecord(decode(payload, RawRecordEnvelope.class));
  }

  /**
   * Fetch many records by ID in windows of {@link #MAX_CONCURRENT_GETS}. Chunking bounds in-flight
   * requests (mirrors the Kotlin chunked-window design). Preserves caller-supplied order.
   */
  public List<Record> get(List<String> recordIds) {
    if (recordIds.isEmpty()) {
      return List.of();
    }
    List<Record> collected = new ArrayList<>(recordIds.size());
    try (ExecutorService executor = Executors.newVirtualThreadPerTaskExecutor()) {
      for (int start = 0; start < recordIds.size(); start += MAX_CONCURRENT_GETS) {
        List<String> window =
            recordIds.subList(start, Math.min(start + MAX_CONCURRENT_GETS, recordIds.size()));
        List<Callable<Record>> tasks =
            window.stream().<Callable<Record>>map(id -> () -> get(id)).toList();
        try {
          List<Future<Record>> futures = executor.invokeAll(tasks);
          for (Future<Record> future : futures) {
            collected.add(future.get());
          }
        } catch (InterruptedException e) {
          Thread.currentThread().interrupt();
          throw new AirtableException.Network(e);
        } catch (ExecutionException e) {
          if (e.getCause() instanceof AirtableException airtable) {
            throw airtable;
          }
          throw new AirtableException.Decoding(e.getCause());
        }
      }
    }
    return collected;
  }

  /** Fetch all records (default query). */
  public List<Record> get() {
    return get(new AirtableQuery());
  }

  /**
   * Fetch records matching the query. Paginates internally using Airtable's {@code offset} token;
   * returns the full merged list.
   */
  public List<Record> get(AirtableQuery query) {
    List<Record> collected = new ArrayList<>();
    String offset = null;
    do {
      String payload =
          offset != null ? fetchWithOffset(query, offset) : client.listRecords(tableId, query);
      RawListResponse envelope = decode(payload, RawListResponse.class);
      if (envelope.records() != null) {
        for (RawRecordEnvelope raw : envelope.records()) {
          collected.add(toRecord(raw));
        }
      }
      offset = envelope.offset() != null && !envelope.offset().isEmpty() ? envelope.offset() : null;
    } while (offset != null);
    return collected;
  }

  /**
   * Refetch with an {@code offset} continuation token appended to the query parameters. {@link
   * AirtableQuery} intentionally doesn't model offset (it's a server-driven continuation, not a
   * user-facing parameter), so it is spliced in at the URL layer.
   */
  private String fetchWithOffset(AirtableQuery query, String offset) {
    List<AirtableQuery.Param> params = new ArrayList<>(query.toParameters());
    params.add(new AirtableQuery.Param("offset", offset));
    return client.send("GET", client.tableUrl(tableId, params), null);
  }

  // ---- create ----

  /** Create one record and return the populated envelope. */
  public Record create(Fields fields) {
    return create(fields, false);
  }

  /**
   * Create one record with optional {@code typecast}. When {@code typecast} is true Airtable
   * coerces string inputs to the cell's type (parsing dates/numbers, creating missing select
   * options).
   */
  public Record create(Fields fields, boolean typecast) {
    List<Record> created = createBatch(List.of(fields), typecast);
    if (created.isEmpty()) {
      throw new AirtableException.Api("UNEXPECTED_RESPONSE", "create returned no records");
    }
    return created.get(0);
  }

  /** Create many records. Chunks into Airtable's 10-per-call batch limit. */
  public List<Record> create(List<Fields> fields) {
    return create(fields, false);
  }

  /** Create many records with optional {@code typecast}. Chunks into Airtable's batch limit. */
  public List<Record> create(List<Fields> fields, boolean typecast) {
    List<Record> collected = new ArrayList<>(fields.size());
    for (int start = 0; start < fields.size(); start += BATCH_SIZE) {
      collected.addAll(
          createBatch(
              fields.subList(start, Math.min(start + BATCH_SIZE, fields.size())), typecast));
    }
    return collected;
  }

  private List<Record> createBatch(List<Fields> fields, boolean typecast) {
    if (fields.isEmpty()) {
      return List.of();
    }
    ObjectNode body = AirtableJson.MAPPER.createObjectNode();
    ArrayNode records = body.putArray("records");
    for (Fields f : fields) {
      ObjectNode record = records.addObject();
      ObjectNode fieldsNode = record.putObject("fields");
      f.toMap().forEach(fieldsNode::set);
    }
    // Airtable echoes the inserted records back keyed by field ID, matching
    // `returnFieldsByFieldId=true` on list/get requests.
    body.put("returnFieldsByFieldId", true);
    // Only emit `typecast` when opted in; Airtable's server-side default is false.
    if (typecast) {
      body.put("typecast", true);
    }
    String payload = client.createRecords(tableId, body.toString());
    RawListResponse envelope = decode(payload, RawListResponse.class);
    List<Record> created = new ArrayList<>();
    if (envelope.records() != null) {
      for (RawRecordEnvelope raw : envelope.records()) {
        created.add(toRecord(raw));
      }
    }
    return created;
  }

  // ---- update ----

  /** Update a single record's field values. */
  public Record update(String recordId, Fields fields) {
    return update(recordId, fields, false);
  }

  /** Update a single record's field values with optional {@code typecast}. */
  public Record update(String recordId, Fields fields, boolean typecast) {
    List<Record> updated = updateBatch(List.of(new Update(recordId, fields)), typecast);
    if (updated.isEmpty()) {
      throw new AirtableException.Api("UNEXPECTED_RESPONSE", "update returned no records");
    }
    return updated.get(0);
  }

  /** Update many records. Chunks into Airtable's 10-per-call batch limit. */
  public List<Record> update(List<Update> updates) {
    return update(updates, false);
  }

  /** Update many records with optional {@code typecast}. Chunks into Airtable's batch limit. */
  public List<Record> update(List<Update> updates, boolean typecast) {
    List<Record> collected = new ArrayList<>(updates.size());
    for (int start = 0; start < updates.size(); start += BATCH_SIZE) {
      collected.addAll(
          updateBatch(
              updates.subList(start, Math.min(start + BATCH_SIZE, updates.size())), typecast));
    }
    return collected;
  }

  private List<Record> updateBatch(List<Update> updates, boolean typecast) {
    if (updates.isEmpty()) {
      return List.of();
    }
    ObjectNode body = AirtableJson.MAPPER.createObjectNode();
    ArrayNode records = body.putArray("records");
    for (Update update : updates) {
      ObjectNode record = records.addObject();
      record.put("id", update.id());
      ObjectNode fieldsNode = record.putObject("fields");
      update.fields().toMap().forEach(fieldsNode::set);
    }
    body.put("returnFieldsByFieldId", true);
    if (typecast) {
      body.put("typecast", true);
    }
    String payload = client.updateRecords(tableId, body.toString());
    RawListResponse envelope = decode(payload, RawListResponse.class);
    List<Record> updated = new ArrayList<>();
    if (envelope.records() != null) {
      for (RawRecordEnvelope raw : envelope.records()) {
        updated.add(toRecord(raw));
      }
    }
    return updated;
  }

  // ---- delete ----

  /** Delete one record by ID. */
  public void delete(String recordId) {
    client.deleteRecord(tableId, recordId);
  }

  /** Delete many records. Chunks into Airtable's 10-per-call batch limit. */
  public void delete(List<String> recordIds) {
    for (int start = 0; start < recordIds.size(); start += BATCH_SIZE) {
      client.deleteRecords(
          tableId, recordIds.subList(start, Math.min(start + BATCH_SIZE, recordIds.size())));
    }
  }
}

File: static/kotlin/DictTable.kt
// ==========================================
// MyAirtable static runtime.
// ==========================================

package myairtable

import io.ktor.http.HttpMethod
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import java.time.Instant

/**
 * Raw-dict record accessor for an Airtable table. Forwards all I/O to the
 * [AirtableClient]. Values flow through as [Fields] (dual ID/name-keyed
 * `Map<String, JsonElement>`), bypassing typed `{Table}Model` decoding.
 *
 * Paired with `OrmTable<T>` (K-F4) — both share the same underlying client.
 * Generated `{Table}Table.dict` exposes this accessor.
 */
class DictTable(
    val tableId: String,
    val nameToId: Map<String, String> = emptyMap(),
    private val client: AirtableClient,
) {
    companion object {
        /** Airtable's hard cap for create / update / delete batches. */
        const val BATCH_SIZE: Int = 10

        /**
         * Max in-flight requests for the multi-ID [get] fan-out. Kept just
         * under Airtable's 5-requests-per-second limit so large ID lists
         * don't trigger a synchronized 429 storm.
         */
        const val MAX_CONCURRENT_GETS: Int = 4
    }

    // region Wire envelopes

    @Serializable
    private data class RawRecordEnvelope(
        val id: RecordId,
        @Serializable(with = AirtableInstantSerializer::class)
        val createdTime: Instant? = null,
        val fields: Map<String, JsonElement> = emptyMap(),
    )

    @Serializable
    private data class RawListResponse(
        val records: List<RawRecordEnvelope> = emptyList(),
        val offset: String? = null,
    )

    /** A fully-loaded record: id + createdTime + [Fields]. */
    data class Record(
        val id: RecordId,
        val createdTime: Instant?,
        val fields: Fields,
    )

    // endregion

    private val json = AirtableJson.instance

    private fun toRecord(raw: RawRecordEnvelope): Record =
        Record(
            id = raw.id,
            createdTime = raw.createdTime,
            fields = Fields(raw.fields, nameToId = nameToId),
        )

    private inline fun <T> decoding(block: () -> T): T =
        try {
            block()
        } catch (e: AirtableException) {
            throw e
        } catch (e: Exception) {
            throw AirtableException.Decoding(e)
        }

    // region get (read)

    /** Fetch one record by ID. */
    suspend fun get(recordId: String): Record {
        val payload = client.getRecord(tableId, recordId)
        val envelope = decoding { json.decodeFromString<RawRecordEnvelope>(payload) }
        return toRecord(envelope)
    }

    /**
     * Fetch many records by ID in windows of [MAX_CONCURRENT_GETS]. Chunking
     * bounds both in-flight requests and live coroutines (a semaphore alone
     * would still suspend one coroutine per ID up-front — PR #21 review).
     * Preserves caller-supplied order.
     */
    suspend fun get(recordIds: List<String>): List<Record> {
        if (recordIds.isEmpty()) return emptyList()
        return recordIds.chunked(MAX_CONCURRENT_GETS).flatMap { window ->
            coroutineScope { window.map { id -> async { get(id) } }.awaitAll() }
        }
    }

    /**
     * Fetch records matching the query. Paginates internally using Airtable's
     * `offset` token; returns the full merged list.
     */
    suspend fun get(query: AirtableQuery = AirtableQuery()): List<Record> {
        val collected = mutableListOf<Record>()
        var offset: String? = null
        do {
            val payload =
                if (offset != null) {
                    fetchWithOffset(query, offset)
                } else {
                    client.listRecords(tableId, query)
                }
            val envelope = decoding { json.decodeFromString<RawListResponse>(payload) }
            collected.addAll(envelope.records.map(::toRecord))
            offset = envelope.offset?.takeIf { it.isNotEmpty() }
        } while (offset != null)
        return collected
    }

    /**
     * Refetch with an `offset` continuation token appended to the query
     * parameters. [AirtableQuery] intentionally doesn't model offset (it's a
     * server-driven continuation, not a user-facing parameter), so it is
     * spliced in at the URL layer.
     */
    private suspend fun fetchWithOffset(
        query: AirtableQuery,
        offset: String,
    ): String {
        val params = query.toParameters() + ("offset" to offset)
        return client.send(HttpMethod.Get, client.tableUrl(tableId, params))
    }

    // endregion

    // region create

    /** Create one record and return the populated envelope. */
    suspend fun create(fields: Fields): Record =
        createBatch(listOf(fields)).firstOrNull()
            ?: throw AirtableException.Api("UNEXPECTED_RESPONSE", "create returned no records")

    /** Create many records. Chunks into Airtable's 10-per-call batch limit. */
    suspend fun create(fields: List<Fields>): List<Record> = fields.chunked(BATCH_SIZE).flatMap { createBatch(it) }

    private suspend fun createBatch(fields: List<Fields>): List<Record> {
        if (fields.isEmpty()) return emptyList()
        // `typecast` is intentionally omitted — matches the Rust / Py / TS
        // behavior (Airtable defaults it to false).
        val body =
            buildJsonObject {
                put(
                    "records",
                    buildJsonArray {
                        for (f in fields) {
                            add(buildJsonObject { put("fields", JsonObject(f.toMap())) })
                        }
                    },
                )
                // Airtable echoes the inserted records back keyed by field ID,
                // matching `returnFieldsByFieldId=true` on list/get requests.
                put("returnFieldsByFieldId", JsonPrimitive(true))
            }
        val payload = client.createRecords(tableId, body.toString())
        val envelope = decoding { json.decodeFromString<RawListResponse>(payload) }
        return envelope.records.map(::toRecord)
    }

    // endregion

    // region update

    /** Update a single record's field values. */
    suspend fun update(
        recordId: String,
        fields: Fields,
    ): Record =
        updateBatch(listOf(recordId to fields)).firstOrNull()
            ?: throw AirtableException.Api("UNEXPECTED_RESPONSE", "update returned no records")

    /** Update many records. Chunks into Airtable's 10-per-call batch limit. */
    suspend fun update(updates: List<Pair<String, Fields>>): List<Record> = updates.chunked(BATCH_SIZE).flatMap { updateBatch(it) }

    private suspend fun updateBatch(updates: List<Pair<String, Fields>>): List<Record> {
        if (updates.isEmpty()) return emptyList()
        val body =
            buildJsonObject {
                put(
                    "records",
                    buildJsonArray {
                        for ((id, f) in updates) {
                            add(
                                buildJsonObject {
                                    put("id", JsonPrimitive(id))
                                    put("fields", JsonObject(f.toMap()))
                                },
                            )
                        }
                    },
                )
                put("returnFieldsByFieldId", JsonPrimitive(true))
            }
        val payload = client.updateRecords(tableId, body.toString())
        val envelope = decoding { json.decodeFromString<RawListResponse>(payload) }
        return envelope.records.map(::toRecord)
    }

    // endregion

    // region delete

    /** Delete one record by ID. */
    suspend fun delete(recordId: String) {
        client.deleteRecord(tableId, recordId)
    }

    /** Delete many records. Chunks into Airtable's 10-per-call batch limit. */
    suspend fun delete(recordIds: List<String>) {
        for (chunk in recordIds.chunked(BATCH_SIZE)) {
            client.deleteRecords(tableId, chunk)
        }
    }

    // endregion
}
